### ISSUE TITLE
Images are not being resized!

### ISSUE THREAD

**Issue by Redani** [apache/incubator-pagespeed-mod#1197]
Hello everybody,
I've already posted some question about mod_pagespeed but never got an answer, so hope i'll get some replies for this issue that i'm currently facing...

I have enabled ModPagespeedEnableFilters rewrite_images, the images are being renamed with the right width and height (/img/150x40xname.png.pagespeed.ic.HASH.png) but it's not resized, it keeps it's original size :/

Do you please have any idea what am I missing?

Thanks

**Comment by jmarantz** [comment]
Where did you post your earlier message.  This group or another one?

Are you saying that if you download 50x40xname.png.pagespeed.ic.HASH.png,
and then run "file 50x40xname.png.pagespeed.ic.HASH.png" it does not tell
you that the image is 50x40?

Could you provide an example URL?

-Josh

**Comment by Redani** [comment]
Wow, thank you very much for the very fast reply :)
Actually I have just check my previous post, you already replied but didn't get any notification :/

I'll ask the same questions here, but before, let's deal with the resizing problem ;)
here is the url of the my website: o p t i g u r a . com (sorry for the spelling, just to prevent the link from being indexed), you can notice that the logo is 150x40 and the original is 300x80, and I don't know why it's not being resized!

here is what I use in the VHOST:

```
<IfModule pagespeed_module>
    ModPagespeed on
    AddOutputFilterByType MOD_PAGESPEED_OUTPUT_FILTER text/html
    ModPagespeedFileCachePath /var/cache/mod_pagespeed/
    ModPagespeedLoadFromFile "https://www.site.com/" "/path/root/"
    ModPagespeedLoadFromFile "https://www.site.com/js/" "/path/root/js/"
    ModPagespeedLoadFromFile "https://www.site.com/css/" "/path/root/css/"
    ModPagespeedLoadFromFile "https://www.site.com/img/" "/path/root/img/"

    ModPagespeedEnableFilters combine_css
    #ModPagespeedEnableFilters combine_javascript
    ModPagespeedEnableFilters extend_cache
    ModPagespeedEnableFilters rewrite_css
    ModPagespeedEnableFilters rewrite_images
    #ModPagespeedEnableFilters rewrite_javascript
    ModPagespeedEnableFilters rewrite_style_attributes_with_url
    ModPagespeedSharedMemoryLocks on

    ModPagespeedMessageBufferSize 100000
</IfModule>
```

Regarding my previous questions, here they are:

> I've just installed pagespeed and i'm kindda impressed how powerful it is, but i have some questions coz i still need to do so many adjustment to make my website completely compatible with this module!!
> 
> 1- I had in my htaccess a rule to force trailing slash:
> RewriteCond %{REQUEST_URI} /+[^.]+$
> RewriteRule ^(.+[^/])$ %{REQUEST_URI}/ [R=301,L]
> 
> but it was returning a 404 error, i had to comment these 2 lines, here is the error: GET https://www.domain.com/mod_pagespeed_beacon/?url=https%3A%2F%2Fwww.domain.com%2Fcategory%2F 404 (Not Found)
> 
> 1.1-Do you know why??

**fixed: by adding `RewriteCond %{REQUEST_URI} !.pagespeed.` or maybe you have a better solution?**

> 1.2- what are the beacon for??
> 
> 2- regarding combining css files, i see this in my element inspector: /inc/A.js,,_slippry,,_slippry.css+css,,_global.css,Mcc.z90cDa6Eiu.css.pagespeed.cf.PgY3-68Bqq.css
> 
> 2.1- what is A.js is doing here?
> 2.2- is it normal the 2 "empty commas"?
> 
> 3- is there a way to minify the html output code using pagespeed?
> 
> 4- in some images, i see onload="pagespeed.CriticalImages.checkImageForCriticality(this);", what is that for?
> 
> Thanks a lot for your help
> Cheers

**Comment by jmarantz** [comment]
On Wed, Dec 2, 2015 at 7:47 PM, Redani notifications@github.com wrote:

> Wow, thank you very much for the very fast reply :)
> Actually I have just check my previous post, you already replied but
> didn't get any notification :/
> 
> Was this the github issue report in ngx_pagespeed?  I am still learning
> github myself.  I'm not sure what the story is with notifications.
> 
> I'll ask the same questions here, but before, let's deal with the resizing
> problem ;)
> here is the url of the my website: o p t i g u r a . com (sorry for the
> spelling, just to prevent the link from being indexed), you can notice that
> the logo is 150x40 and the original is 300x80, and I don't know why it's
> not being resized!
> 
> You are right.  The resulting file looks resized based on its name, but is
> not actually resized.  This is a bug.  We'll try to reproduce and see
> what's going on.
> 
> 1- I had in my htaccess a rule to force trailing slash:
> RewriteCond %{REQUEST_URI} /+[^.]+$
> RewriteRule ^(.+[^/])$ %{REQUEST_URI}/ [R=301,L]
> 
> but it was returning a 404 error, i had to comment these 2 lines, here is
> the error: GET
> https://www.domain.com/mod_pagespeed_beacon/?url=https%3A%2F%2Fwww.domain.com%2Fcategory%2F
> 404 (Not Found)
> 
> 1.1-Do you know why??
> 
> _fixed: by adding RewriteCond %{REQUEST_URI} !.pagespeed. or maybe you
> have a better solution?_

In general, PageSpeed may not always play well with mod_rewrite.  The
workaround you suggested is probably the right one.

> 1.2- what are the beacon for??
> 
> We use beacons for a few things:
1. Determining which images are critical (above the fold).  We only
   inline critical images, and only lazyload non-critical.
2. Computing critical css for prioritize_critical_css
   https://developers.google.com/speed/pagespeed/module/filter-prioritize-critical-css
3. Measuring page-load time for add_instrumentation
   https://developers.google.com/speed/pagespeed/module/filter-instrumentation-add

2- regarding combining css files, i see this in my element inspector:

> /inc/A.js,,_slippry,,_slippry.css+css,,_global.css,Mcc.z90cDa6Eiu.css.pagespeed.cf.PgY3-68Bqq.css
> 
> 2.1- what is A.js is doing here?
> 
> The "A" is historical, and it would be a bit annoying for us to remove
> it.  There was a time in the past when we would encode the class of browser
> in a prefix in front of all CSS files: "A." meant "archaic" e.g. IE6 and
> IE7.  "W." meant webp-capable (Opera, Chrome, Android), and "I." meant
> "inline-capable", e.g. Firefox.  This is all related what we did to images
> referenced from CSS files.  Later we decided to use a different mechanism
> in our system to distinguish these three cases, and we just made everything
> "A.".
> 
> 2.2- is it normal the 2 "empty commas"?
> 
> TL;DR: Yes.

More detail: that particular URL you gave resulted from two CSS files that
were combined, and then the combination was rewritten to minify it and
optimize any referenced images.  These two operations mean there was a
double-encoding.  One of the source files was /inc/js/slippry/slippry.css,
the other was /inc/css/global.css.  When combining two files in different
directories, we put the combination in the lowest common root, which is
"/inc".  Then "/" characters in the file get escaped.  Our escape-character
is ",", and the escape-sequence for slash is ",_".  The escape sequence for
"," is ",,".  So when the two files were combined the slashes were replaced
like /inc/js,_slippry,_/slippry.css.  And when the combined file was
minified, another round of escaping occurred, giving us ",,_" for each of
the original slashes.

All this is done this way so that if the cache is flushed, we can decode
and recreate the combination from the URL.

3- is there a way to minify the html output code using pagespeed?

> Yes, enables some or all these filters:

collapse_whitespace
https://developers.google.com/speed/pagespeed/module/filter-whitespace-collapse
remove_comments
https://developers.google.com/speed/pagespeed/module/filter-comment-remove

remove_quotes
https://developers.google.com/speed/pagespeed/module/filter-quote-remove

> 4- in some images, i see
> onload="pagespeed.CriticalImages.checkImageForCriticality(this);", what is
> that for?
> 
> For determining which images are above-the fold when clients render your
> web page.

-Josh

> Thanks a lot for your help
> 
> My pleasure!

-Josh

**Comment by jmarantz** [comment]
Bug reported here: https://github.com/pagespeed/mod_pagespeed/issues/1198

Please subscribe to that to track progress.

-Josh

**Comment by Redani** [comment]
Thanks again for your celerity ;)

> Was this the github issue report in ngx_pagespeed?

Yes, it was in that group, which obviously was the wrong one :D 

> You are right.  The resulting file looks resized based on its name, but is
> not actually resized.  This is a bug.  We'll try to reproduce and see
> what's going on.

Well, what is weird is that it works fine for some images in the home page, if you scroll down a little bit, you'll see 3 products, the original size is 200x200px and they are resized to 180x180, but the quality of the image is a little bad!! is that because i have retina screen?
it also works fine in this example page: http://www.modpagespeed.com/rewrite_images.html?ModPagespeed=on&ModPagespeedFilters=rewrite_images

> Yes, enables some or all these filters:
> collapse_whitespace
> ....

Yes, i've seen these filters, but i was more talking about compressing and/or minifying the HTML output, just like minifying the css or js

Thanks for the detailed explanations of the rest of my questions :) I'll read it with more attention tomorrow ;)

One other question plz, are you planing to add support to combine external js and css files? like fontawesome, fonts.googleapis.com, facebook, ... when i run tests on pagespeed, there are always problems with these kind of files :/

Regards

P.S. I've subscribed to the reported bug, looking forward to updates... :)

**Comment by jmarantz** [comment]
On Wed, Dec 2, 2015 at 8:57 PM, Redani notifications@github.com wrote:

> Thanks again for your celerity ;)
> 
> Was this the github issue report in ngx_pagespeed?
> 
> Yes, it was in that group, which obviously was the wrong one :D
> 
> You are right. The resulting file looks resized based on its name, but is
> 
> not actually resized. This is a bug. We'll try to reproduce and see
> what's going on.
> 
> Well, what is weird is that it works fine for some images in the home
> page, if you scroll down a little bit, you'll see 3 products, the original
> size is 200x200px and they are resized to 180x180,
> 
> Right, well we do have tests we run on every commit that show that image
> resizing works generally.  There might be something about that particular
> image that makes us fail to inline it.  We'll debug it.
> 
> but the quality of the image is a little bad!! is that because i have
> retina screen?
> 
> I have good news about that: in our next release we will add better support
> for retina screens by automatically adding srcset
> http://www.w3.org/html/wg/drafts/html/master/semantics.html#attr-img-srcset
> attributes for sized images.  That should be out within the next week or so.
> 
> Yes, enables some or all these filters:
> collapse_whitespace
> ....
> 
> Yes, i've seen these filters, but i was more talking about compressing
> and/or minifying the HTML output, just like minifying the css or js
> 
> What kind of minifying are you thinking about, beyond what we do already?
> Usually we just rely on mod_deflate to compress the responses after we
> minify them.  That seems to be working fine on your site.
> 
> Thanks for the detailed explanations of the rest of my questions :) I'll
> read it with more attention tomorrow ;)
> 
> One other question plz, are you planing to add support to combine external
> js and css files? like fontawesome, fonts.googleapis.com, facebook, ...
> when i run tests on pagespeed, there are always problems with these kind of
> files :/
> 
> See MapProxyDomain
> https://developers.google.com/speed/pagespeed/module/domains#MapProxyDomain
> for
> external domains in general (at least domains that you trust), and see
> inline_google_font_css
> https://developers.google.com/speed/pagespeed/module/filter-css-inline-google-fonts
> for
> Google fonts in particular.

-Josh

**Comment by Redani** [comment]
Hi Josh

> What kind of minifying are you thinking about, beyond what we do already?
> Usually we just rely on mod_deflate to compress the responses after we
> minify them.  That seems to be working fine on your site.

I was talking about transforming something like this:

`````` <html>
  <head>
    <style>
      .blue {
         color: blue;
      }
    </style>
    </head>
  <body>
    <div class="blue">
      Hello, world!
    </div>
  </body>
</html>```

into this:

```<html><head><style>.blue{color:blue;}</style></head><body><div class="blue">Hello, world!</div></body></html>```
``````

**Comment by oschaaf** [comment]
@Redani you might want to check https://developers.google.com/speed/pagespeed/module/filter-whitespace-collapse :-)

**Comment by Redani** [comment]
hi @oschaaf , I know that filter, but it doesn't do what i want :) i'd like to remove also the line breaks...

**Comment by Redani** [comment]
> See MapProxyDomain
> https://developers.google.com/speed/pagespeed/module/domains#MapProxyDomain
> for
> external domains in general (at least domains that you trust), and see
> inline_google_font_css
> https://developers.google.com/speed/pagespeed/module/filter-css-inline-google-fonts
> for
> Google fonts in particular

I have tried filter-css-inline-google-fonts but didn't work, I also enabled "ModPagespeedFetchHttps enable" but nothing changed in my page...

I also tried to use MapProxyDomain but i'm having trouble to understand it.
what should I put in target_domain/subdir (mysite.com/subdir ?), what subdir is for?
same for origin_domain/subdir and also [rewrite_domain/subdir]
does this mean that the scripts will be copied and hosted locally?
and what if they update their scripts, will it be taken into consideration?

Thanks

**Comment by oschaaf** [comment]
@Redani 

MapProxyDomain will proxy and optimize external domains. As for considering changes to assets on proxied domains: the expiry of the proxied assets is respected, and changes will be reflected accordingly.
I'm not able to explain its parameterization any better then the docs, but perhaps an example may help here?:

```
pagespeed MapProxyDomain http://www.mydomain.com/externalimages/ http://externalimages.com/
```

**Comment by jmarantz** [comment]
HI -- this has really turned into a discussion-group thing and not a bug
thing.

RE minifying HTML; we don't feel comfortable removing all whitespace; it
can change appearance of the page.  It's safer to change multiple
consecutive whitespace characters into a single one.  It's actually not
100% safe even to do that.  We look for 'pre' tags and avoid whitespace
collapsing in them, but whitespace can be significant due to CSS settings,
which we can't be sure of when serving hTML.  This is why
collapse_whitespace is not on by default.

RE font-inlining: see
https://groups.google.com/forum/#!topic/ngx-pagespeed-discuss/AJr1-ChMPxY
-- I think you need to set
   ModPagespeedCssInlineMaxBytes 5000
The size of the google font loader css changed after we released the filter
and so you need that directive.

**Comment by Redani** [comment]
Hi,

@oschaaf i have used this: `ModPagespeedMapProxyDomain https://www.mysite.com/tmp/ https://www.google-analytics.com/` and also `ModPagespeedMapProxyDomain https://www.mysite.com/tmp/ https://www.google-analytics.com/analytics.js` but didn't see any change, the file analytics is still the same!

@jmarantz I added the following directives, but still nothing happens regarding google fonts:

```
ModPagespeedEnableFilters inline_google_font_css
ModPagespeedFetchHttps enable
ModPagespeedCssInlineMaxBytes 5000
```

Regards

**Comment by oschaaf** [comment]
@Redani @jmarantz pointed this out to me, and my own measurements on IISpeed.com back this:  I do not think proxying analytics.js is going to help much: the file is very common, and you would bust any cached versions of it that your visitors have already if you pull it into your own domain.
Next to that, the hosting of analytics.js will probably outperform the hosting of your website. 

Having said that, It looks like MapProxyDomain on https does not seem to work seamlessly - I had to combine MapProxyDomain with MapOriginDomain and MapRewriteDomain to get it to work when I experimented with it: https://disqus.com/home/discussion/iispeed/optimize_your_pagespeed_insights_score_when_using_tracking_and_social_media_scripts/#comment-2138127663

I will create a separate issue later to track what is up with MapProxyDomain on https.

**Comment by Redani** [comment]
Hey guys, hope you're doing well!

I'm facing a little small problem with a jQuery plugin i'm using: http://jquery.eisbehr.de/lazy/ to display an image I use `<img src="/img/blank.jpg" data-src="/img/prods/my-real-image.jpg" alt="" />`, and it's the image "blank.jpg" which is being rewritten by pagespeed in stead of the other :/ is there a work around to fix this?
Thanks a lot

**Comment by jmarantz** [comment]
Hi Redani,

In the future if you are just asking questions can you do it on the
-discuss list rather than a bug?

in this case the answer is quite easy: check out 'url valued attributes':
https://developers.google.com/speed/pagespeed/module/domains#url-valued-attributes

Also consider using mod_pagespeed's own lazyload images filter:
https://developers.google.com/speed/pagespeed/module/filter-lazyload-images

In particular, this filter helps the server learn which images are not
critical to the above-the-fold view, and only lazy-loads those ones.

-Josh

**Comment by Redani** [comment]
> In the future if you are just asking questions can you do it on the discuss list rather than a bug?
> Yes sure, i will, sorry about that ;)

Just wanted to let you know that I have added `ModPagespeedUrlValuedAttribute img data-src image` and it's working well, thanks a lot ;)

I dont find the discuss list, could you plz send me the link?

P.S. any updates regarding the issue of resizing images?

**Comment by oschaaf** [comment]
@Redani mod-pagespeed-discuss can be found over here: https://groups.google.com/forum/#!forum/mod-pagespeed-discuss

**Comment by Redani** [comment]
Hi, i'd like to ask if you have any updates about the image resize issue. I see there is a new release: 1.10.32.2, is it official and stable?
